### PR TITLE
Add type hints

### PR DESCRIPTION
**Pull request by zupo**
Also upgrade dev env to latest versions.

### CODE REVIEW

File: .mypy/type_coverage.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    )
+    parser.add_argument(
+        "coverage",
+        type=float,
+        metavar="<coverage>",
+        help="Minimum required type coverage.",
+    )
+    parser.add_argument(
+        "file",
+        type=str,
```
Comment (sayanarijit) [inline .mypy/type_coverage.py]: Better way `type=argparse.FileType('r')`

Comment (zupo) [reply .mypy/type_coverage.py]: I'm not familiar with argparse.FileType, why is it better?

Comment (sayanarijit) [reply .mypy/type_coverage.py]: It will eliminate lines 29-31. You won't have to verify if file exists and then open it.

Comment (zupo) [reply .mypy/type_coverage.py]: Sweet!

**Pull request merged**